###############################
###############################
#Version v0.6
# This R script is the second module of the ngsLCA toolkit, which processes and visualizes the lca files generated by the first module
# Developed and tested in R version 3.6.1
# Wiki can be found at https://github.com/miwipe/ngsLCA
# Please report bugs and/or suggestions to wyc661217@gmail.com
###############################
###############################


###############################
#install and request R packages
###############################
packs <- c("vegan","gplots","circlize","reshape","RColorBrewer","rioja","readr","BiocManager","devtools","tidyr","ggplot2")
biocondpacks <- c("ComplexHeatmap")
libs <- .libPaths()
installedPacks <- rownames(installed.packages(lib.loc=libs))
needPacks <- packs[!packs%in%installedPacks]
needPacks2 <- biocondpacks[!biocondpacks%in%installedPacks]

if(length(c(needPacks,needPacks2))>0){
  afile=tempfile()
  cat('\n\n\t->  You need to install packages. Run below command in R:\n\n')
  if(length(needPacks)>0){
    cat("install.packages(",file=afile)
    for(i in 1:length(needPacks)){
      cat("\"",needPacks[i],"\"",file=afile,append=T)
      if(i<length(needPacks))
        cat(",",file=afile,append=T)
    }
    cat(")\n",file=afile,append=T)
  }
  if(length(needPacks2)>0){
    cat("BiocManager::install(c(",file=afile,append=T)
    for(i in 1:length(needPacks2)){
      cat("\"",needPacks2[i],"\"",file=afile,append=T)
      if(i<length(needPacks2))
        cat(",",file=afile,append=T)
    }
    cat("))\n",file=afile,append=T)
  }
  tmp<-gsub(" ","",readLines(afile))
  if(length(tmp)>0)
    cat(" \t",tmp[1],"\n")
  if(length(tmp)>1)
    cat("\n\t",tmp[2],"\n")
  unlink(afile)
  quit()
}


cat("\n\n\t-> loading R packages \n\n")
tmp<-lapply(c(packs,biocondpacks), require, character.only = TRUE, quietly = T)


###############################
#0
#Parameters read-in
###############################
rm(list=(ls()))
options(warn=-1)

l<-commandArgs(TRUE)

getArgs<-function(x,l){
  unlist(strsplit(grep(paste("^",x,"=",sep=""),l,val=T),"="))[2]
}


Args<-function(l,args){
  if(! all(sapply(strsplit(l,"="),function(x)x[1])%in%names(args))){
    cat("Error -> ",l[!sapply(strsplit(l,"="),function(x)x[1])%in%names(args)]," is not a valid argument\n")
    q("no")
  }
  arguments<-list()
  for(a in names(args))
    arguments[[a]]<-getArgs(a,l)
  
  if(any(!names(args)%in%names(arguments)&sapply(args,is.null))){
    cat("Error -> ",names(args)[!names(args)%in%names(arguments)&sapply(args,is.null)]," is not optional!\n")
    q("no")
  }
  for(a in names(args))
    if(is.null(arguments[[a]]))
      arguments[[a]]<-args[[match(a,names(args))]]
  
  arguments
}

print.args<-function(args,des){
  if(missing(des)){
    des<-as.list(rep("",length(args)))
    names(des)<-names(args)
  }
  cat("->  See https://github.com/miwipe/ngsLCA for wiki\n")
  cat("->  Needed arguments:\n")
  mapply(function(x)cat("\t",x,":",des[[x]],"\n"),cbind(names(args)[sapply(args,is.null)]))
  cat("->  Optional arguments (defaults):\n")
  mapply(function(x)cat("\t",x," (",args[[x]],")",":",des[[x]],"\n"),cbind(names(args)[!sapply(args,is.null)]))
  q("no")
}

#NULL is an non-optional argument; NA is an optional argument; others are the default arguments
args<-list(
  path = NULL,
  path_blank = NA,
  output = "R_result",
  task = "pre-process,filter,de-contamination,group,rank,count,megan,krona,heatmap,barplot,rarefy",
  metadata = NA,
  threshold.1 = 2,
  threshold.2 = 0,
  threshold.3 = 5,
  threshold.1_blank = NA,
  threshold.2_blank = NA,
  threshold.3_blank = NA,
  remove.taxa = NA,
  remove.sample = NA,
  group.name = "Viruses,Archaea,Bacteria,Fungi,Viridiplantae,Metazoa",
  rank.name = "species,genus,family",
  threshold.perGroup = 0,
  top.abundance = 30,
  NMDS_trymax = 1000
)

##if no argument are given this will print the arguments description
des<-list(
  path = "working directory containing the lca files that will be processed\n",
  path_blank = "working directory containing the lca files for all blank controls\n",
  output = "name of the output folder under the working directory\n",
  task = "functions that will be performed, other options: NMDS, stratplot\n",
  metadata = "path to your metadata, an example at https://github.com/miwipe/ngsLCA\n",
  threshold.1 = "minimum reads number required for confirming a taxon in each sample\n",
  threshold.2 = "minimum reads percentage (to the total reads number of the sample) required for confirming a taxon in each sample, ranging from 0 to 1\n",
  threshold.3 = "minimum summed reads number across all samples required for confirming a taxon in the combined taxonomic profile\n",
  threshold.1_blank = "same as threshold.1, but for filtering the blank controls; will using the value of threshold.1 if not specified\n",
  threshold.2_blank = "same as threshold.2, but for filtering the blank controls; will using the value of threshold.2 if not specified\n",
  threshold.3_blank = "same as threshold.3, but for filtering the blank controls; will using the value of threshold.3 if not specified\n",
  remove.taxa = "a list of NCBI taxaID indicating the taxa that will be removed from final results\n",
  remove.sample = "a list of file names indicating the samples that will be removed from the final results\n",
  group.name = "higher taxonomic units that will be used for grouping the taxa, need to be as the scientific names in NCBI taxonomy\n",
  rank.name = "taxonomic ranks that will be used for clustering the taxa\n",
  threshold.perGroup = "minimum reads percentage (to the total reads number of each group) required for confirming a taxon in each sample, ranging from 0 to 1\n",
  top.abundance = "how many most abundant taxa will be illustrated in figs\n",
  NMDS_trymax = "maximum numbers of random starts in search of convergent solutions for NMDS\n"
)


#Load arguments
if(length(l)==0) print.args(args,des)
attach(Args(l,args))
args <- commandArgs(TRUE)
if(length(args)==0){
  cat(" Arguments: output prefix\n")
  q("no")
}

#process arguments
if (!grepl('/$', path)) {
  path = paste(path,"/",sep="")
}


if (!is.na(path_blank)) {
  if (!grepl('/$', path_blank)) {
    path_blank = paste(path_blank,"/",sep="")
  }
}

if (!is.na(task)) {
  func = strsplit(task,",")[[1]]
}

if (!is.na(remove.taxa)) {
  remove.taxa = as.numeric(strsplit(remove.taxa,",")[[1]])
}

if (!is.na(remove.sample)) {
  remove.sample = strsplit(remove.sample,",")[[1]]
}

if (!is.na(group.name)) {
  group.name = strsplit(group.name,",")[[1]]
}

if (!is.na(rank.name)) {
  rank.name = strsplit(rank.name,",")[[1]]
}

thr1=as.numeric(threshold.1)
thr2=as.numeric(threshold.2)
thr3=as.numeric(threshold.3)
thrG=as.numeric(threshold.perGroup)
top.abundance=as.numeric(top.abundance)
NMDS_trymax=as.numeric(NMDS_trymax)


if (is.na(threshold.1_blank)) {
  threshold.1_blank = thr1
}

if (is.na(threshold.2_blank)) {
  threshold.2_blank = thr2
}

if (is.na(threshold.3_blank)) {
  threshold.3_blank = thr3
}

thr1b=as.numeric(threshold.1_blank)
thr2b=as.numeric(threshold.2_blank)
thr3b=as.numeric(threshold.3_blank)


cat("\t-> arguments will be applied:\n")
cat("\t\tpath = ", path,"\n")
cat("\t\tpath_blank = ", path_blank,"\n")
cat("\t\toutput = ", output,"\n")
cat("\t\ttask = ", func,"\n")
cat("\t\tmetadata = ", metadata,"\n")
cat("\t\tthreshold.1 = ", thr1,"\n")
cat("\t\tthreshold.2 = ", thr2,"\n")
cat("\t\tthreshold.3 = ", thr3,"\n")
cat("\t\tthreshold.1_blank = ", thr1b,"\n")
cat("\t\tthreshold.2_blank = ", thr2b,"\n")
cat("\t\tthreshold.3_blank = ", thr3b,"\n")
cat("\t\tremove.taxa = ", remove.taxa,"\n")
cat("\t\tremove.sample = ", remove.sample,"\n")
cat("\t\tgroup.name = ", group.name,"\n")
cat("\t\trank.name = ", rank.name,"\n")
cat("\t\tthreshold.perGroup = ", thrG,"\n")
cat("\t\ttop.abundance = ", top.abundance,"\n")
cat("\t\tNMDS_trymax = ", NMDS_trymax,"\n")



######################################
#1
#pre-process
######################################
#function for lca file read-in, and replace the file name with metadata if supplied
ReadIn = function(FileList,path){
  #creat a black dataframe
  X1 = data.frame(taxa=character(), stringsAsFactors=F)

  #read in
  for (i in 1:length(FileList)) {
    X2.1 =  read.csv(paste(path, FileList[i], sep=""), header=F, sep="\t", stringsAsFactors=F, fill=T, 
                     col.names = paste0("V",seq_len(60)),comment.char = "#")
    
    if(dim(X2.1)[1]>0){
      
      X2.2 = data.frame(taxa=X2.1[,2],
                        count=rep(1, dim(X2.1)[1]),
                        stringsAsFactors = F)
      
      X2.3 = aggregate(X2.2[,2]~X2.2[,1], data=X2.2, FUN=sum)
      colnames(X2.3) = c("taxa",sub(".lca","",FileList[i]))
      X1 = merge(X1, X2.3, by="taxa", all=T)
      
    }
  }
  
  X1[is.na(X1)] = 0
  
  #Removes samples in the sample removing list
  if (!is.na(remove.sample)){
    remove.sample = sub(".lca", "", remove.sample)
    X1 = X1[,!(colnames(X1) %in% remove.sample)]
  }
  
  #replcing the file names with metadata
  if(!is.na(metadata)){
    Mdata = read.csv(metadata, quote="", stringsAsFactors=F, header=F,sep = "\t")
    Mdata$V1 = sub(".lca","",Mdata$V1)
    Mdata = Mdata[order(Mdata$V3),]
    
    N = 0
    for (i in 2:dim(X1)[2]) {
      if (length(which(colnames(X1)[i] %in% Mdata$V1))>0) {
        colnames(X1)[i] = Mdata$V2[which(Mdata$V1 %in% colnames(X1)[i])]
        N = c(N,i)
      }
    }
    
    N = N[-1]
    M = which(!2:dim(X1)[2] %in% N) + 1
    X1 = X1[,c(1,N,M)]
    
  }
  
  #Removes taxa in the taxa.re list
  if (!is.na(remove.taxa)) {
    TAXAID = as.numeric(sapply(strsplit(X1$taxa, split = ":"),function(x) x[1]))
    if (length(which(TAXAID %in% remove.taxa)) >0 ) {
      X1 = X1[-which(TAXAID %in% remove.taxa),]
    }
  }
  return(X1)
}

#read-in
if ("pre-process" %in% func) {
  
  cat("\n\n\t-> Data pre-process, it might take some time depending on file number and size\n\n")
  dir.create(paste(path, output, sep=""))
  dir.create(paste(path, output, "/intermediate", sep=""))
  
  file.list1 = dir(path, pattern = ".lca$")
  if(length(file.list1)==0){
    cat("\n\n\t-> ERROR: no lca files found in the appointed working directory\n\n")
    q("no")
  }
  
  if(is.na(metadata)){
    cat("\n\n\t-> Metadata file not supplied and file names will be illustrated \n\n")
  } else {
    cat("\n\n\t-> Metadata supplied and will be used \n\n")
  }
  
  DF1 = ReadIn(FileList=file.list1, path=path)
  if (dim(DF1)[1]>0) {
    write.table(DF1, file = paste(path, output, "/intermediate/", "taxa_profile_v1.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
  }else{
    cat("\n\n\t-> ERROR: no alignments appear in the input lca files\n\n")
    q("no")
  }
  
  
  #blanks
  if (is.na(path_blank)) {
    file.list2 = NA
  }else{
    file.list2 = dir(path_blank, pattern = ".lca$")
  }
  
  if(is.na(file.list2)){
    cat("\n\n\t-> The path to blank controls not supplied or no lca files in it, contamination list will not be generated \n\n")
  } else{
    
    CON = ReadIn(FileList=file.list2, path=path_blank)
    
    if (dim(CON)[1]>0) {
      cat("\n\n\t-> The path to blank controls supplied, a contamination list will be generated \n\n")
      write.table(CON, file = paste(path, output, "/intermediate/", "contamination_list_v1.txt", sep=""), quote=F, 
                  row.names=F, col.names=T, sep="\t")
    }else{
      cat("\n\n\t-> No alignments appear in the supplied blacnk control lca files \n\n")
    }
  }
  
  #generating the taxa branch
  PATH = read.csv(paste(path, file.list1[1], sep=""), header=F, sep="\t", stringsAsFactors=F, 
                  fill=T, col.names = paste0("V",seq_len(60)),comment.char = "#")
  PATH = PATH[,-1]
  
  if (length(file.list1)>1) {
    for (i in 2:length(file.list1)) {
      PATH.1 = read.csv(paste(path, file.list1[i], sep=""), header=F, sep="\t", stringsAsFactors=F, 
                        fill=T, col.names = paste0("V",seq_len(60)),comment.char = "#")
      PATH.1 = PATH.1[,-1]
      PATH = rbind(PATH,PATH.1)
      PATH = PATH[!duplicated(PATH[,1]),]
    }
  }else{
    PATH = PATH[!duplicated(PATH[,1]),]
  }
  write.table(PATH, file = paste(path, output, "/intermediate/", "taxa_branch.txt", sep=""), quote=F, 
              row.names=F, col.names=T, sep="\t")
}



######################################
#2
#filter
######################################
if ("filter" %in% func) {
  
  cat("\n\n\t-> filter the taxonomic profile\n\n")
  
  if (length(dir(paste(path, output, "/intermediate/", sep=""),pattern =  "taxa_profile_v1.txt")) == 0) {
    cat("\n\n\t-> Required file not found in the appointed directory, please run the 'pre-process' first \n\n")
    q("no")
  }
  
  DF1 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v1.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  DF2.1 = as.data.frame(sapply(DF1[,-1], as.numeric))
  colnames(DF2.1) = colnames(DF1)[-1]
  DF2.1[DF2.1 < thr1] = 0
  DF2.1$taxa  = DF1$taxa
  DF2.1 = DF2.1[,c(dim(DF2.1)[2],2:dim(DF2.1)[2]-1)]
  
  if (dim(DF2.1)[2]>2) {
    DF2.1 = DF2.1[which(rowSums(DF2.1[,-1]) !=0),]
  }else{
    DF2.1 = DF2.1[which(DF2.1[,2] !=0),]
  }
  
  if (dim(DF2.1)[1]>0) {
    write.table(DF2.1, file = paste(path, output, "/intermediate/", "taxa_profile_v2.1.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
  }else{
    cat("\n\n\t-> All taxa removed by the threshold.1, try assign a lower value \n\n")
    q("no")
  }
  
  
  DF2.2 = DF2.1
  for (i in 2:dim(DF2.2)[2]) {
    if (sum(DF2.2[,i])>0) {
      DF2.2[(DF2.2[,i] < sum(DF2.2[,i])*thr2),i] = 0
    }
  }
  
  if (dim(DF2.2)[1]>0) {
    write.table(DF2.2, file = paste(path, output, "/intermediate/", "taxa_profile_v2.2.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
  }else{
    cat("\n\n\t-> All taxa removed by the threshold.2, try assign a lower value \n\n")
    q("no")
  }
  
  
  
  if (dim(DF2.2)[2]>2) {
    DF2.3 = DF2.2[which(rowSums(DF2.2[,-1]) >= thr3),]
  }else{
    DF2.3 = DF2.2[which(DF2.2[,2] >= thr3),]
  }
  
  if (dim(DF2.3)[1]>0) {
    write.table(DF2.3, file = paste(path, output, "/intermediate/", "taxa_profile_v2.3.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
  }else{
    cat("\n\n\t-> All taxa removed by the threshold.3, try assign a lower value \n\n")
    q("no")
  }
  
  #filter blank controls
  if (length(dir(paste(path, output,"/intermediate/", sep=""), pattern = "contamination_list_v1.txt"))>0) {
    
    DF1 = read.delim(paste(path, output, "/intermediate/", "contamination_list_v1.txt", sep=""), 
                     stringsAsFactors=FALSE,header = T,check.names = F)
    
    DF2.1 = as.data.frame(sapply(DF1[,-1], as.numeric))
    colnames(DF2.1) = colnames(DF1)[-1]
    DF2.1[DF2.1 < thr1b] = 0
    DF2.1$taxa  = DF1$taxa
    DF2.1 = DF2.1[,c(dim(DF2.1)[2],2:dim(DF2.1)[2]-1)]
    
    if (dim(DF2.1)[2]>2) {
      DF2.1 = DF2.1[which(rowSums(DF2.1[,-1]) !=0),]
    }else{
      DF2.1 = DF2.1[which(DF2.1[,2] !=0),]
    }
    
    write.table(DF2.1, file = paste(path, output, "/intermediate/", "contamination_list_v2.1.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
    
    
    DF2.2 = DF2.1
    for (i in 2:dim(DF2.2)[2]) {
      if (sum(DF2.2[,i])>0) {
        DF2.2[(DF2.2[,i] < sum(DF2.2[,i])*thr2b),i] = 0
      }
    }
    
    write.table(DF2.2, file = paste(path, output, "/intermediate/", "contamination_list_v2.2.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
    
    
    if (dim(DF2.2)[2]>2) {
      DF2.3 = DF2.2[which(rowSums(DF2.2[,-1]) >= thr3b),]
    }else{
      DF2.3 = DF2.2[which(DF2.2[,2] >= thr3b),]
    }
    
    write.table(DF2.3, file = paste(path, output, "/intermediate/", "contamination_list_v2.3.txt", sep=""), quote=F, 
                row.names=F, col.names=T, sep="\t")
    
  }
}



######################################
#3
#de-contamination
######################################
if ("de-contamination" %in% func) {
  
  cat("\n\n\t-> remove taxa detected in the contamination list\n\n")
  
  if (length(dir(paste(path, output, "/intermediate/", sep=""),pattern="taxa_profile_v2.3.txt")) == 0) {
    cat("\n\n\t-> Required file not found in the appointed directory, please run the 'pre-process' and 'filter' first \n\n")
    q("no")
  }
  
  DF1 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v2.3.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)

  if (length(dir(paste(path, output, "/intermediate/", sep=""),pattern="contamination_list_v2.3.txt")) == 0) {
    cat("\n\n\t-> No contamination list found in the appointed directory, de-contamination function denied.\n\tMake sure the right path to blank controls supplied, and the 'pre-process' and 'filter' functions performed \n\n")
    DF2 = DF1
  }else{
    CON = read.delim(paste(path, output, "/intermediate/", "contamination_list_v2.3.txt", sep=""), 
                     stringsAsFactors=FALSE,header = T)
    if (length(which(DF1$taxa %in% CON$taxa))>0) {
      DF2 = DF1[-which(DF1$taxa %in% CON$taxa),]
    }else{
      DF2 = DF1
    }
  }
  
  write.table(DF2, file = paste(path, output, "/intermediate/", "taxa_profile_v3.txt", sep=""), quote=F, 
              row.names=F, col.names=T, sep="\t")
  
  dir.create(paste(path, output, "/taxonomic_profiles", sep=""))
  DF3 = DF2 %>% separate(taxa, sep=":", c("taxa_ID","taxa_name","taxonomic_rank"))
  write.table(DF3, file = paste(path, output, "/taxonomic_profiles/", "complete_profile.txt", sep=""), quote=F, 
              row.names=F, col.names=T, sep="\t")
}


if (!"de-contamination" %in% func & length(dir(paste(path, output, "/intermediate/", sep=""),pattern =  "taxa_profile_v3.txt")) == 0) {
  
  if (length(dir(paste(path, output, "/intermediate/", sep=""),pattern =  "taxa_profile_v2.3.txt")) == 0) {
    cat("\n\n\t-> Required file not found in the appointed directory, please run the 'pre-process' and 'filter' first \n\n")
    q("no")
  }
  
  DF1 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v2.3.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  
  write.table(DF1, file = paste(path, output, "/intermediate/", "taxa_profile_v3.txt", sep=""), quote=F, 
              row.names=F, col.names=T, sep="\t")
  
  dir.create(paste(path, output, "/taxonomic_profiles", sep=""))
  DF2 = DF1 %>% separate(taxa, sep=":", c("taxa_ID","taxa_name","taxonomic_rank"))
  write.table(DF2, file = paste(path, output, "/taxonomic_profiles/", "complete_profile.txt", sep=""), quote=F, 
              row.names=F, col.names=T, sep="\t")
}


######################################
#4
#group
######################################
#group function
GroupTaxa = function(DF, TaxaUnit){
  
  L1 = NULL
  L1[group.name] = list(NULL)
  
  for (i in 1:dim(DF)[1]) {
    for (j in 1:length(TaxaUnit)) {
      if (length(grep(TaxaUnit[j], tax.branch[which(tax.branch[,1] %in%  DF[i,1]),])) != 0) {
        L1[[j]] = c(L1[[j]],i)
      }
    }
  }
  
  name = data.frame(taxa=names(L1))
  
  for (i in 1:length(L1)) {
    if (length(L1[[i]]) != 0) {
      
      DF1 = DF[L1[[i]],]
      
      if (dim(DF1)[2]>2) {
        for (j in 1:dim(DF1)[1]) {
          DF1[j, which(DF1[j,-1] <= colSums(DF1[,-1]) * thrG)+1] = 0
        }
      }else{
        for (j in 1:dim(DF1)[1]) {
          DF1[j, which(DF1[j,-1] <= sum(DF1[,-1]) * thrG)+1] = 0
        }
      }
      
      DF2 = DF1[rowSums(as.data.frame(DF1[,-1])) != 0,]
      write.table(DF2, file = paste(path, output, "/intermediate/taxa_groups/", name$taxa[i], ".txt", sep=""), quote=F, 
                  row.names=F, col.names=T, sep="\t")
      
      DF2.1 = DF2 %>% separate(taxa, sep=":", c("taxa_ID","taxa_name","taxonomic_rank"))
      write.table(DF2.1, file = paste(path, output, "/taxonomic_profiles/taxa_groups/", name$taxa[i], ".txt", sep=""), 
                  quote=F,row.names=F, col.names=T, sep="\t")
    }
  }
}
#group
if ("group"%in%func) {
  
  cat("\n\n\t-> group taxa\n\n")
  
  dir.create(paste(path,output, "/taxonomic_profiles/taxa_groups", sep=""))
  dir.create(paste(path,output, "/intermediate/taxa_groups", sep=""))
  
  tax.branch = read.csv(paste(path, output,"/intermediate/taxa_branch.txt", sep=""), 
                        sep="\t", quote="", stringsAsFactors=F)
  DF1 = read.csv(paste(path, output,"/intermediate/", "taxa_profile_v3.txt", sep=""),
                 sep="\t", quote="", check.names=F,stringsAsFactors=F)
  
  GroupTaxa(DF = DF1, TaxaUnit = group.name)
}



######################################
#5
#rank
######################################
#Function for clustering taxa into user-defined taxa ranks
Taxa.cluster = function(DF, OutName){
  
  DF1 = data.frame(matrix(vector(), 0, dim(DF)[2]+1, dimnames=list(c(), c("taxa",colnames(DF)))),stringsAsFactors=F)
  colnames(DF1)[2:dim(DF1)[2]] = colnames(DF)
  
  j=1
  for (i in 1:dim(DF)[1]) {
    V1 = grep(Taxa.L, tax.branch[which(tax.branch[,1] %in% rownames(DF)[i]),])
    if (length(V1)>1) {
      V1=V1[1]
    }
    if (length(V1) !=0) {
      DF1[j,] = c(tax.branch[which(tax.branch[,1] %in% rownames(DF)[i]),V1], as.numeric(DF[i,1:dim(DF)[2]]))
      j = j+1
    }
  }
  
  for (i in 2:dim(DF1)[2]) {
    DF1[,i] = as.numeric(DF1[,i])
  }
  
  if (dim(DF1)[1] != 0) {
    DF2 = aggregate(. ~ taxa, data = DF1, sum)
    DF2 = DF2 %>% separate(taxa, sep=":", c("taxa_ID","taxa_name","taxonomic_rank"))
    write.table(DF2, quote=F, row.names=F, col.names=T, sep="\t",
                file = paste(path, output, "/taxonomic_profiles/taxa_ranks/", OutName, "_",taxa.level, ".txt", sep=""))
  }else{
    cat("\n\n\t-> for", OutName, ":::no taxa clustered into", taxa.level, "\n\n")
  }
}

#cluster
if ("rank"%in%func) {
  
  dir.create(paste(path, output, "/taxonomic_profiles/taxa_ranks", sep=""))
  cat("\n\n\t-> cluster taxa into taxonomic ranks\n\n")
  
  DF = read.csv(paste(path, output, "/intermediate/", "taxa_profile_v3.txt", sep=""),sep="\t", 
                quote="", check.names=F,stringsAsFactors=F)
  tax.branch = read.csv(paste(path, output, "/intermediate/", "taxa_branch.txt", sep=""), 
                        sep="\t",quote="", stringsAsFactors=F)
  
  rownames(DF) = DF$taxa
  DF1 = as.data.frame(DF[,-1])
  colnames(DF1) = colnames(DF)[-1]
  rownames(DF1) = rownames(DF)
  
  for (taxa.level in rank.name) {
    Taxa.L = paste(":",taxa.level,sep="")
    X2 = Taxa.cluster(DF = DF1, OutName = "all_taxa")
  }
  
  #cluster on groups
  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_groups",sep=""), pattern = ".txt"))>0) {
    
    file.list = dir(paste(path, output, "/intermediate/taxa_groups/", sep=""), pattern = ".txt") #list files
    
    for (i in 1:length(file.list)) {
      
      DF = read.csv(paste(path, output, "/intermediate/taxa_groups/", file.list[i], sep=""),sep="\t", 
                    quote="", check.names=F,stringsAsFactors=F)
      rownames(DF) = DF$taxa
      DF1 = as.data.frame(DF[,-1])
      colnames(DF1) = colnames(DF)[-1]
      rownames(DF1) = rownames(DF)
      
      OutName = sub(".txt", "", file.list[i])
      
      for (taxa.level in rank.name) {
        Taxa.L = paste(":",taxa.level,sep="")
        X2 = Taxa.cluster(DF = DF1, OutName = OutName)
      }
      
    }
  }
}




######################################
#6
#count
######################################
if ("count"%in%func) {
  
  cat("\n\n\t-> counting reads and taxa numbers\n\n")
  dir.create(paste(path, output, "/counts", sep=""))
  dir.create(paste(path, output, "/counts/readsNO", sep=""))
  dir.create(paste(path, output, "/counts/taxaNO", sep=""))
  
  DF1 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v1.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  DF2.1 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v2.1.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  DF2.2 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v2.2.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  DF2.3 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v2.3.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  DF3 = read.delim(paste(path, output, "/intermediate/", "taxa_profile_v3.txt", sep=""), 
                   stringsAsFactors=FALSE,header = T,check.names = F)
  
  ReadNO = data.frame(sample=c(colnames(DF1)[-1]),
                      original_Reads=colSums(DF1[,-1]),
                      passed_threshold.1=colSums(DF2.1[,-1]),
                      passed_threshold.2=colSums(DF2.2[,-1]),
                      passed_threshold.3=colSums(DF2.3[,-1]),
                      contaminate_removed=colSums(DF3[,-1]),
                      stringsAsFactors = F)
  
  for (i in 2:dim(DF1)[2]) {
    DF1[which(DF1[,i] > 0),i] = 1
    DF2.1[which(DF2.1[,i] > 0),i] = 1
    DF2.2[which(DF2.2[,i] > 0),i] = 1
    DF2.3[which(DF2.3[,i] > 0),i] = 1
    DF3[which(DF3[,i] > 0),i] = 1
  }
  
  TaxaNO =  data.frame(sample=c(colnames(DF1)[-1]),
                       taxa_in_original_lca=colSums(DF1[,-1]),
                       passed_threshold.1=colSums(DF2.1[,-1]),
                       passed_threshold.2=colSums(DF2.2[,-1]),
                       passed_threshold.3=colSums(DF2.3[,-1]),
                       contaminate_removed=colSums(DF3[,-1]),
                       stringsAsFactors = F)
  
  if (length(dir(paste(path, output, "/intermediate/taxa_groups",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/intermediate/taxa_groups/", sep=""), pattern = ".txt")
    
    for (i in 1:length(file.list)) {
      
      DF = read.csv(paste(path, output, "/intermediate/taxa_groups/", file.list[i], sep=""),sep="\t", 
                    quote="", check.names=F,stringsAsFactors=F)
      ReadNO$new = colSums(DF[,-1])
      colnames(ReadNO)[dim(ReadNO)[2]] = sub(".txt", "", file.list[i])
      
      for (j in 2:dim(DF)[2]) {
        DF[which(DF[,j] > 0),j] = 1
      }
      
      TaxaNO$new = colSums(DF[,-1])
      colnames(TaxaNO)[dim(TaxaNO)[2]] = sub(".txt", "", file.list[i])
    }
  }
  
  write.table(ReadNO,quote=F, row.names=F, col.names=T, sep="\t",
              file = paste(path, output, "/counts/","ReadNO.txt", sep=""))
  write.table(TaxaNO,quote=F, row.names=F, col.names=T, sep="\t",
              file = paste(path, output, "/counts/","TaxaNO.txt", sep=""))
  
  
  for (i in 1:dim(ReadNO)[1]) {
    
    D1 = data.frame(step=colnames(ReadNO)[-1],
                    number=as.numeric(ReadNO[i,-1]),
                    stringsAsFactors = F)
    D1$step = factor(D1$step, as.character(D1$step),ordered = T)
    
    D2 = data.frame(step=colnames(TaxaNO)[-1],
                    number=as.numeric(TaxaNO[i,-1]),
                    stringsAsFactors = F)
    D2$step = factor(D2$step, as.character(D2$step),ordered = T)
    
    Name = ReadNO$sample[i]
    
    
    p1=ggplot(data=D1, aes(x=step, y=number)) +
      geom_bar(stat="identity")+
      ggtitle(ReadNO$sample[i])+
      labs(y = "Reads number") +
      theme(axis.title.x = element_text(size = 13),
            axis.title.y = element_text(size = 13),
            axis.text.x = element_text(size = 11, colour = "black" ,angle = 40),
            axis.text.y = element_text(size = 11, colour = "black"),
            legend.text = element_text(size = 15),
            panel.grid.major = element_blank(),
            panel.background = element_blank(),
            panel.border = element_blank(),
            axis.line.x = element_line(colour = "black",size = 0.5, linetype = "solid"),
            axis.line.y = element_line(colour = "black",size = 0.5, linetype = "solid"),
            axis.ticks.length = unit(0.2, "cm"))
    
    ggsave(plot=p1, height=8, width=6, dpi=300,
           filename=paste(path, output, "/counts/readsNO/",ReadNO$sample[i],"_ReadNO.pdf", sep=""), 
           useDingbats=FALSE)
    
    p2=ggplot(data=D2, aes(x=step, y=number)) +
      geom_bar(stat="identity")+
      ggtitle(ReadNO$sample[i])+
      labs(y = "Reads number") +
      theme(axis.title.x = element_text(size = 13),
            axis.title.y = element_text(size = 13),
            axis.text.x = element_text(size = 11, colour = "black" ,angle = 40),
            axis.text.y = element_text(size = 11, colour = "black"),
            legend.text = element_text(size = 15),
            panel.grid.major = element_blank(),
            panel.background = element_blank(),
            panel.border = element_blank(),
            axis.line.x = element_line(colour = "black",size = 0.5, linetype = "solid"),
            axis.line.y = element_line(colour = "black",size = 0.5, linetype = "solid"),
            axis.ticks.length = unit(0.2, "cm"))
    
    ggsave(plot=p2, height=8, width=6, dpi=300,
           filename=paste(path, output, "/counts/taxaNO/",ReadNO$sample[i],"_TaxaNO.pdf", sep=""), 
           useDingbats=FALSE)
  }
}



######################################
#7
#Genarate files for MEGAN
######################################
if ("megan"%in%func) {
  
  cat("\n\n\t-> generate MEGAN files\n\n")
  dir.create(paste(path, output, "/megan", sep=""))
  
  X1 = read.csv(paste(path, output, "/taxonomic_profiles/complete_profile.txt", sep=""), 
                sep="\t", quote="",check.names=F,stringsAsFactors=F)
  
  X1 = X1[,-c(2,3)]
  colnames(X1)[1] = "#dataset"
  write.table(X1, file = paste(path, output,"/megan/all_taxa.txt", sep=""), quote=F, row.names=F, col.names=T, sep=",")
  
  
  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_groups",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_groups",sep=""), pattern = ".txt")
    for (i in 1:length(file.list)) {
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_groups/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X1 = X1[,-c(2,3)]
      colnames(X1)[1] = "#dataset"
      write.table(X1, file = paste(path, output, "/megan/", file.list[i], sep=""), quote=F, row.names=F, col.names=T, sep=",")
    }
  }
  
  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt")
    for (i in 1:length(file.list)) {
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_ranks/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X1 = X1[,-c(2,3)]
      colnames(X1)[1] = "#dataset"
      write.table(X1, file = paste(path, output, "/megan/", file.list[i], sep=""), quote=F, row.names=F, col.names=T, sep=",")
    }
  }
}



######################################
#8
#Genarate files for krona
######################################
if ("krona"%in%func) {
  
  cat("\n\n\t-> generate krona files\n\n")
  dir.create(paste(path, output, "/krona", sep=""))
  
  tax.branch = read.csv(paste(path, output, "/intermediate/", "taxa_branch.txt", sep=""), 
                        sep="\t", quote="", stringsAsFactors=F)
  krona.branch = data.frame(matrix(nrow = dim(tax.branch)[1], ncol = 2), stringsAsFactors=F)
  
  for (i in 1:dim(tax.branch)[1]) {
    
    krona.branch[i,1] = paste(strsplit(tax.branch[i,1],":")[[1]][1:2], collapse = ":")
    
    Taxa.B = strsplit(krona.branch[i,1],":")[[1]][2]
    
    for (j in 2:dim(tax.branch)[2]) {
      if (!is.na(tax.branch[i,j])){
        Taxa.B = paste(strsplit(tax.branch[i,j],":")[[1]][2], Taxa.B, collapse = "/t")
      }
    }
    
    krona.branch[i,2] = Taxa.B
  }
  
  X1 = read.csv(paste(path, output, "/intermediate/", "taxa_profile_v3.txt", sep=""), 
                quote="",sep="\t",check.names=F,stringsAsFactors=F)
  
  for (i in 2:dim(X1)[2]) {
    
    X2 = data.frame(X1[,i], X1[,1], stringsAsFactors=F)
    X2 = X2[X2[,1] != 0,]
    X2 = X2[order(-X2[,1]),]
    
    if (dim(X2)[1]>0) {
      
      if (dim(X2)[1] >top.abundance) {
        X2 = X2[1:top.abundance,]
      }
      
      for (j in 1:dim(X2)[1]) {
        X2[j,2] = paste(strsplit(X2[j,2],":")[[1]][c(1,2)], collapse = ":")
      }
      
      
      for (j in 1:dim(X2)[1]){
        X2[j,2] = krona.branch[which(krona.branch[,1] == X2[j,2], arr.ind=T), 2]
      }
      
      write.table(X2, file = paste(path, output, "/krona/krona_", colnames(X1)[i] ,".txt", sep=""), 
                  quote=F, row.names=F, col.names=F, sep="\t")
      
    }
  }
}



######################################
#9.1
#heatmap
######################################
#function for heatmap
HeatMap = function(DF){
  #transfer data into percentage and subset
  for (i in 1:dim(DF)[2]) {
    if(sum(DF[,i]) != 0){
      DF[,i] = DF[,i]/sum(DF[,i])
    }
  }
  DF$sum = rowSums(DF)
  DF = DF[order(-DF$sum),]
  
  if (dim(DF)[1] > top.abundance) {
    DF = DF[1:top.abundance,-dim(DF)[2]]
  } else{
    DF = DF[,-dim(DF)[2]]
  }
  
  DF = as.matrix(DF)
  if (length(which(colSums(DF) == 0))>0) {
    F1 = Heatmap(DF, column_dend_height = unit(1.5, "cm"), row_dend_width = unit(3, "cm"), show_row_names = T, show_column_names = T,
                 row_names_gp = gpar(cex=0.8), column_names_gp = gpar(cex=0.8), cluster_rows = T, cluster_columns= F, 
                 clustering_distance_rows = "pearson", clustering_distance_columns = "pearson",
                 col = colorRamp2(c(0, 0.001, 0.01, 0.03, 0.06, 0.2), 
                                  c("white", "cornflowerblue", "yellow", "#FD8D3C","#E31A1C","#B10026")),
                 rect_gp = gpar(col = "gray12", lty = 1, lwd = 0.5),
                 heatmap_legend_param = list(title = "abundance", title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), 
                                             legend_height = unit(10, "cm"), color_bar = "continous"))
  }else{
    F1 = Heatmap(DF, column_dend_height = unit(1.5, "cm"), row_dend_width = unit(3, "cm"), show_row_names = T, show_column_names = T,
                 row_names_gp = gpar(cex=0.8), column_names_gp = gpar(cex=0.8), cluster_rows = T, cluster_columns= T, 
                 clustering_distance_rows = "pearson",clustering_distance_columns = "pearson",
                 col = colorRamp2(c(0, 0.001, 0.01, 0.03, 0.06, 0.2), 
                                  c("white", "cornflowerblue", "yellow", "#FD8D3C","#E31A1C","#B10026")),
                 rect_gp = gpar(col = "gray12", lty = 1, lwd = 0.5),
                 heatmap_legend_param = list(title = "abundance", title_gp = gpar(fontsize = 8), labels_gp = gpar(fontsize = 8), 
                                             legend_height = unit(10, "cm"), color_bar = "continous"))
  }

  return(F1)
}

dataPrep = function(DF){
  
  if (length(which(duplicated(DF[,2])))>0) {
    N=which(DF[,2] %in% DF[which(duplicated(DF[,2])),2])
    DF[N,2] = paste(DF[N,2],DF[N,3],sep="_")
    if (length(which(duplicated(DF[,2])))>0) {
      N=which(DF[,2] %in% DF[which(duplicated(DF[,2])),2])
      DF[N,2] = paste(DF[N,1],DF[N,2],sep="_")
    }
  }

  row.names(DF) = DF[,2]
  DF2 = as.data.frame(DF[,-c(1:3)])
  colnames(DF2) = colnames(DF)[-c(1:3)]
  rownames(DF2) = rownames(DF)
  
  return(DF2)
}

if ("heatmap"%in%func){
  
  cat("\n\n\t-> generate heatmap\n\n")
  dir.create(paste(path, output, "/heatmap", sep=""))
  
  X1 = read.csv(paste(path, output, "/taxonomic_profiles/complete_profile.txt", sep=""),
                sep="\t", quote="", check.names=F,stringsAsFactors=F)
  X2 = dataPrep(X1)
  
  if (dim(X2)[2]>1) {
    pdf(paste(path, output, "/heatmap/complete_profile_heatmap.pdf", sep=""), width=8, height=13)
    print({HeatMap(X2)})
    dev.off() 
  }else{
    cat("\n\n\t-> only one sample detected, heatmap function turn off\n\n")
  }
  
  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt")
    
    for (i in 1:length(file.list)) {
      
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_ranks/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X2 = dataPrep(X1)
      
      Name = sub(".txt", "", file.list[i])
      
      if (dim(X2)[2]>1) {
        pdf(paste(path, output, "/heatmap/",Name,"_heatmap.pdf", sep=""), width=8, height=13)
        print({HeatMap(X2)})
        dev.off() 
      } else {
        cat(paste("\n\n\t-> only one sample detected in ", file.list[i],", heatmap function turn off for this file \n\n",sep = ""))
      }
    }
  }
}



######################################
#9.2
#barplot
######################################
if ("barplot"%in%func){
  
  cat("\n\n\t-> generating barplot \n\n")
  dir.create(paste(path, output, "/barplot", sep=""))
  
  X1 = read.csv(paste(path, output, "/taxonomic_profiles/complete_profile.txt", sep=""),
                sep="\t", quote="", check.names=F,stringsAsFactors=F)
  X2 = dataPrep(X1)
  
  for (i in 1:dim(X2)[2]) {
    if (sum(X2[,i])>0) {
      X2[,i] = X2[,i]/sum(X2[,i])
    }
  }
  sample_list = colnames(X2)
  X2$taxa = rownames(X2)
  X2 = melt(X2,"taxa")
  X2$variable = factor(X2$variable,sample_list,ordered = T)
  
  p0=ggplot(X2, aes(x=variable, y=value,fill =taxa)) +
    geom_bar(stat="identity")+
    labs(y="Proportion (%)")+
    theme(axis.title.x = element_blank(),
          plot.title = element_text(size=17, face="bold",hjust = 0.5),
          legend.title = element_blank(),
          legend.text = element_text(size = 15),
          axis.title.y = element_text(size = 17),
          axis.text.x = element_text(size = 15, angle = 15, colour = "black"),
          axis.text.y = element_text(size = 15, colour = "black"),
          panel.grid.major = element_blank(),
          panel.background = element_blank(),
          panel.border = element_rect(colour = "black",fill = "transparent",size=1.5),
          legend.position="bottom")
  
  ggsave(plot=p0, height=10, width=15, dpi=500,
         filename=paste(path, output, "/barplot/all_taxa_barplot.pdf", sep=""), 
         useDingbats=FALSE)
  

  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt")
    
    for (i in 1:length(file.list)) {
      
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_ranks/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X2 = dataPrep(X1)
      
      for (j in 1:dim(X2)[2]) {
        if (sum(X2[,j])>0) {
          X2[,j] = X2[,j]/sum(X2[,j])
        }
      }
      sample_list = colnames(X2)
      X2$taxa = rownames(X2)
      X2 = melt(X2,"taxa")
      X2$variable = factor(X2$variable,sample_list,ordered = T)

      Name = sub(".txt", "", file.list[i])
      
      p1 = ggplot(X2, aes(x=variable, y=value,fill =taxa)) +
        geom_bar(stat="identity")+
        labs(y="Proportion (%)")+
        theme(axis.title.x = element_blank(),
              plot.title = element_text(size=17, face="bold",hjust = 0.5),
              legend.title = element_blank(),
              legend.text = element_text(size = 15),
              axis.title.y = element_text(size = 17),
              axis.text.x = element_text(size = 15, angle = 15, colour = "black"),
              axis.text.y = element_text(size = 15, colour = "black"),
              panel.grid.major = element_blank(),
              panel.background = element_blank(),
              panel.border = element_rect(colour = "black",fill = "transparent",size=1.5),
              legend.position="bottom")
     
      ggsave(plot=p1, height=10, width=15, dpi=500,
             filename=paste(path, output, "/barplot/", Name, "_barplot.pdf",sep=""), 
             useDingbats=FALSE)
    }
  }
}


######################################
#10
#rarefaction
######################################
#Function for rarefaction
Raref = function(DF, OutName){
  
  if (pmin(dim(DF)[1],dim(DF)[2]) >= 2) {
    
    DF1 = t(DF)
    S = specnumber(DF1) # observed number of taxa
    raremax = min(rowSums(DF1))
    Srare = rarefy(DF1, raremax)
    #
    pdf(paste(path, output, "/rarefaction/", OutName, "_rarefy_scaling.pdf", sep=""), width=12, height=8) 
    plot(S, Srare, xlab = "Observed No. of taxa", ylab = "Rarefied No. of taxa")
    abline(0, 1)
    dev.off()
    #
    pdf(paste(path, output, "/rarefaction/", OutName, "_rarefy_rarecurve.pdf", sep=""), width=12, height=8) 
    rarecurve(DF1, step = 20, sample = raremax, col = "blue", cex = 1.0)
    dev.off()
  } else {
    cat("\n\n\t-> for", OutName, ":::Number of samples or taxa is less than 2, insufficient for rarefaction \n\n")
  }
}
#rarefy
if ("rarefy"%in%func){
  
  cat("\n\n\t-> random rarefaction \n\n")
  dir.create(paste(path, output, "/rarefaction", sep=""))
  
  X1 = read.csv(paste(path, output, "/taxonomic_profiles/complete_profile.txt", sep=""),
                sep="\t", quote="", check.names=F,stringsAsFactors=F)
  X2 = dataPrep(X1)
  
  Raref(DF=X2, OutName="all_taxa")
  
  
  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt")
    
    for (i in 1:length(file.list)) {
      
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_ranks/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X2 = dataPrep(X1)
      
      Name = sub(".txt", "", file.list[i])
      Raref(DF=X2, OutName=Name)
    }
  }
}


######################################
#11
#NMDS
######################################
#function for NMDS
NMDSc = function(DF, OutName){
  
  NMDS.dimention = as.integer((pmin(dim(DF)[1],dim(DF)[2])-1)/2)+1 #caculate the NMDS dimerntion
  
  if (NMDS.dimention >= 2) {
    NMDS = metaMDS(DF, k=NMDS.dimention, trymax=NMDS_trymax) #NMDS cluster
    saveRDS(NMDS, file = paste(path, output, "/NMDS/", OutName, "_NMDS_data.rda", sep=""))
    #output NMDS_stressplot figure
    pdf(paste(path, output, "/NMDS/", OutName, "_NMDS_stressplot.pdf", sep=""), width=12, height=12) 
    stressplot(NMDS)
    dev.off()
    #output NMDS figure
    X3 = as.data.frame(NMDS$points)[,c(1,2)]
    X4 = as.data.frame(NMDS$species)[,c(1,2)]
    
    p1 = ggplot(X4, aes(MDS1, MDS2))+
      geom_point(size=4,colour="#7F7F7F")+
      geom_text(aes(label = rownames(X4)),colour="black",size=5)+
      labs(title=paste(OutName, "_NMDS", sep=""),
           y="NMDS2", x="NMDS1") +
      theme(axis.title.x = element_text(size = 17),
            title = element_text(size = 19, face="bold"),
            axis.title.y = element_text(size = 17),
            axis.text.x = element_text(size = 13, colour = "black"),
            axis.text.y = element_text(size = 13, colour = "black"),
            legend.title = element_blank(),
            legend.text = element_text(size = 17),
            panel.grid.major = element_blank(),
            panel.background = element_blank(),
            axis.line = element_line(colour = "black",size=0.5),
            panel.border = element_rect(colour = "black",fill = "transparent",size=1.8))+
      geom_density_2d()+
      geom_point(data = X3, 
                 mapping = aes(x = MDS1, y = MDS2),size=1.5,pch=17,colour="#FFE1FF")
    
    ggsave(plot=p1,height=8,width=10,dpi=500, filename=paste(path, output, "/NMDS/", OutName, "_NMDS.pdf", sep=""), useDingbats=FALSE)
  } else {
    cat("\n\n\t-> for", OutName, ":::Number of samples or taxa is less than 2, insufficient for NMDS \n\n")
  }
}


if ("NMDS"%in%func) {
  
  cat("\n\n\t-> NMDS, will take some time depending on input file size and the NMDS_trymax\n\n")
  dir.create(paste(path, output, "/NMDS", sep=""))
  
  # NMDS on taxa profiles for all samples
  X1 = read.csv(paste(path, output, "/taxonomic_profiles/complete_profile.txt", sep=""),
                sep="\t", quote="", check.names=F,stringsAsFactors=F)
  X2 = dataPrep(X1)
  
  NMDSc(DF = X2, OutName = "all_taxa")
  
  
  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt")
    
    for (i in 1:length(file.list)) {
      
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_ranks/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X2 = dataPrep(X1)
      
      Name = sub(".txt", "", file.list[i])
      NMDSc(DF=X2, OutName=Name)
    }
  }
}


######################################
#12
#stratplot
######################################
if ("stratplot"%in%func) {
  
  cat("\n\n\t-> generate stratplot\n\n")
  dir.create(paste(path, output, "/stratplot", sep=""))
  
  X1 = read.csv(paste(path, output, "/taxonomic_profiles/complete_profile.txt", sep=""),
                sep="\t", quote="", check.names=F,stringsAsFactors=F)
  X2 = dataPrep(X1)
  
  for (i in 1:dim(X2)[2]){
    if (sum(X2[,i])>0) {
      X2[,i] = X2[,i]/sum(X2[,i])
    }
  }
  
  if (dim(X2)[1]>top.abundance) {
    X2 = X2[order(rowSums(X2),decreasing = T)[1:top.abundance],]
  }
  
  if (dim(X2)[1]>30) {
    cat("\n\n\t-> top.abundance is too high for valid stratplots; will illustrating the top 30 taxa only\n\n")
    X2 = X2[order(rowSums(X2),decreasing = T)[1:30],]
  }

  X2 = as.data.frame(t(X2),stringsAsFactors = F)
  pdf(paste(path, output, "/stratplot/all_taxa_stratplot.pdf", sep=""), width=15, height=8)
  strat.plot(X2, y.rev=T, plot.line=F, plot.poly=T, plot.bar=T, col.bar="black", 
             col.poly=c(rep("forestgreen",dim(X2)[2])),col.poly.line="black", 
             scale.percent=TRUE, xSpace=0.01, x.pc.lab=TRUE, x.pc.omit0=TRUE, las=2)
  dev.off()

  if (length(dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt"))>0){
    
    file.list = dir(paste(path, output, "/taxonomic_profiles/taxa_ranks",sep=""), pattern = ".txt")
    
    for (i in 1:length(file.list)) {
      
      X1 = read.csv(paste(path, output, "/taxonomic_profiles/taxa_ranks/", file.list[i], sep=""), 
                    sep="\t", quote="",check.names=F,stringsAsFactors=F)
      X2 = dataPrep(X1)
      
      for (j in 1:dim(X2)[2]){
        if (sum(X2[,j])>0) {
          X2[,j] = X2[,j]/sum(X2[,j])
        }
      }
      
      if (dim(X2)[1]>top.abundance) {
        X2 = X2[order(rowSums(X2),decreasing = T)[1:top.abundance],]
      }
      
      X2 = as.data.frame(t(X2),stringsAsFactors = F)
      Name = sub(".txt", "", file.list[i])
      
      pdf(paste(path, output, "/stratplot/",Name,"_stratplot.pdf", sep=""), width=15, height=8)
      p <- strat.plot(X2, plot.bar = TRUE, plot.line = TRUE, plot.poly = F, y.rev=TRUE, col.poly.line='dark green', 
                      col.poly='dark green', col.bar = 'dark green', title="viridiplantae", ylabel="Strata", 
                      srt.xlabel = 45, xSpace = 0.003, lwd.bar = 3)
      dev.off()
    }
  }
}

#End
######################################
######################################

